--- af9035.c	2020-02-01 13:23:53.893666528 +0100
+++ af9035-mod.c	2020-02-01 13:27:52.751655400 +0100
@@ -1215,49 +1215,49 @@
 	struct si2168_config si2168_config;
 	struct i2c_adapter *adapter;
 
-	dev_dbg(&intf->dev, "adap->id=%d\n", adap->id);
+	dev_dbg(&intf->dev, "%s  adap->id=%d\n", __func__, adap->id);
 
-	/* I2C master bus 2 clock speed 366k */
-	ret = af9035_wr_reg(d, 0x00f6a7, I2C_SPEED_366K);
-	if (ret < 0)
-		goto err;
-
-	/* I2C master bus 1,3 clock speed 366k */
-	ret = af9035_wr_reg(d, 0x00f103, I2C_SPEED_366K);
-	if (ret < 0)
-		goto err;
-
-	/* set gpio11 low */
-	ret = af9035_wr_reg_mask(d, 0xd8d4, 0x01, 0x01);
-	if (ret < 0)
-		goto err;
-
-	ret = af9035_wr_reg_mask(d, 0xd8d5, 0x01, 0x01);
-	if (ret < 0)
-		goto err;
-
-	ret = af9035_wr_reg_mask(d, 0xd8d3, 0x01, 0x01);
-	if (ret < 0)
-		goto err;
-
-	/* Tuner enable using gpiot2_en, gpiot2_on and gpiot2_o (reset) */
-	ret = af9035_wr_reg_mask(d, 0xd8b8, 0x01, 0x01);
-	if (ret < 0)
-		goto err;
-
-	ret = af9035_wr_reg_mask(d, 0xd8b9, 0x01, 0x01);
-	if (ret < 0)
-		goto err;
-
-	ret = af9035_wr_reg_mask(d, 0xd8b7, 0x00, 0x01);
-	if (ret < 0)
-		goto err;
-
-	msleep(200);
-
-	ret = af9035_wr_reg_mask(d, 0xd8b7, 0x01, 0x01);
-	if (ret < 0)
-		goto err;
+	/* I2C master bus 2 clock speed 300k */
+  ret = af9035_wr_reg(d, 0x00f6a7, 0x07);
+  if (ret < 0)
+    goto err;
+
+  /* I2C master bus 1,3 clock speed 300k */
+  ret = af9035_wr_reg(d, 0x00f103, 0x07);
+  if (ret < 0)
+    goto err;
+
+  /* set gpio11 low */
+  ret = af9035_wr_reg_mask(d, 0xd8d4, 0x01, 0x01);
+  if (ret < 0)
+    goto err;
+
+  ret = af9035_wr_reg_mask(d, 0xd8d5, 0x01, 0x01);
+  if (ret < 0)
+    goto err;
+
+  ret = af9035_wr_reg_mask(d, 0xd8d3, 0x01, 0x01);
+  if (ret < 0)
+    goto err;
+
+  /* Tuner enable using gpiot2_en, gpiot2_on and gpiot2_o (reset) */
+  ret = af9035_wr_reg_mask(d, 0xd8b8, 0x01, 0x01);
+  if (ret < 0)
+    goto err;
+
+  ret = af9035_wr_reg_mask(d, 0xd8b9, 0x01, 0x01);
+  if (ret < 0)
+    goto err;
+
+  ret = af9035_wr_reg_mask(d, 0xd8b7, 0x00, 0x01);
+  if (ret < 0)
+    goto err;
+
+  msleep(200);
+
+  ret = af9035_wr_reg_mask(d, 0xd8b7, 0x01, 0x01);
+  if (ret < 0)
+    goto err;
 
 	memset(&si2168_config, 0, sizeof(si2168_config));
 	si2168_config.i2c_adapter = &adapter;
